streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -752,54 +750,53 @@ sealed trait ZChannel[-Env, -InErr, -InElem, -InDone, +OutErr, +OutElem, +OutDon
-                       outgoing.offer(ZChannel.failLeftUnit) *>
-                       ZChannel.failUnit
-                 }
-               )
-               .raceFirst(errorSignal.await.interruptible)
-               .forkIn(scope)
+        childScope  <- scope.fork
+        fiberId     <- ZIO.fiberId
+        _ <-
+          pull.flatMap { outElem =>
+            val latch = Promise.unsafe.make[Nothing, Unit](fiberId)(Unsafe)
+            for {
+              _ <- permits
+                     .withPermit(
+                       latch.succeed(()) *> f(outElem)
+                         .foldCauseZIO(
+                           cause =>
+                             failure.update(_ && cause).unless(cause.isInterruptedOnly) *>
+                               errorSignal.succeed(()) *>
+                               outgoing.offer(ZChannel.failLeftUnit),
+                           elem => outgoing.offer(Exit.succeed(elem))
+                         )
+                     )
+                     .interruptible
+                     .forkIn(childScope)
+              _ <- latch.await
+            } yield ()
+          }.forever.interruptible
+            .onError(_.failureOrCause match {
+              case Left(x: Left[OutErr, OutDone]) =>
+                failure.update(_ && Cause.fail(x.value)) *>
+                  outgoing.offer(ZChannel.failLeftUnit)
+              case Left(x: Right[OutErr, OutDone]) =>
+                permits.withPermits(n.toLong)(ZIO.unit).interruptible *>
+                  outgoing.offer(Exit.fail(x.asInstanceOf[Either[Unit, OutDone]]))
+              case Right(cause) =>
+                failure.update(_ && cause).unless(cause.isInterruptedOnly) *>
+                  outgoing.offer(ZChannel.failLeftUnit)
+            })
+            .raceFirst(ZChannel.awaitErrorSignal(childScope, fiberId)(errorSignal))
+            .forkIn(scope)
       } yield {
         lazy val writer: ZChannel[Env1, Any, Any, Any, OutErr1, OutElem2, OutDone] =
           ZChannel.unwrap[Env1, Any, Any, Any, OutErr1, OutElem2, OutDone] {
             outgoing.take.map {
               case s: Exit.Success[OutElem2] => ZChannel.write(s.value) *> writer
               case f: Exit.Failure[Either[Unit, OutDone]] =>
+                def extractFailures = ZChannel.unwrap(failure.get.map(ZChannel.refailCause(_)))
                 f.cause.failureOrCause match {
-                  case Left(_: Left[Unit, OutDone])  => ZChannel.unwrap(failure.get.map(ZChannel.refailCause(_)))
-                  case Left(x: Right[Unit, OutDone]) => ZChannel.succeedNow(x.value)
-                  case Right(cause)                  => ZChannel.refailCause(cause)
+                  case Left(_: Left[Unit, OutDone])    => extractFailures
+                  case Left(x: Right[Unit, OutDone])   => ZChannel.succeedNow(x.value)
+                  case Right(c) if c.isInterruptedOnly => extractFailures
```

Comment:
@regiskuckaertz I noticed that this case was missing here but was present in `mapOutZIOPar`. Was this intentional?

Reply:
no, thank you for fixing 🙏